proposals/memory-specializations/utilities.tex

```diff
@@ -0,0 +1,636 @@
+                                             default_sentinel{}).base();
+\end{codeblock}
+} %% \color{addclr}
+
+\rSec3[uninitialized.copy]{\tcode{uninitialized_copy}}
+{\color{remclr}
+\begin{codeblock}
+template <class InputIterator, class ForwardIterator>
+  ForwardIterator uninitialized_copy(InputIterator first, InputIterator last,
+                                     ForwardIterator result);
+\end{codeblock}
+
+\setcounter{Paras}{1}
+\pnum
+\returns \tcode{result}
+\begin{codeblock}
+template <class InputIterator, class Size, class ForwardIterator>
+  ForwardIterator uninitialized_copy_n(InputIterator first, Size n,
+                                       ForwardIterator result);
+\end{codeblock}
+
+\pnum
+\effects As if by:
+\begin{codeblock}
+        for ( ; n > 0; ++result, (void) ++first, --n) {
+          ::new (static_cast<void*>(addressof(*result)))
+            typename iterator_traits<ForwardIterator>::value_type(*first);
+        }
+\end{codeblock}
+
+\pnum
+\returns \tcode{result}
+} %% \color{remclr}
+
+{\color{addclr}
+\begin{codeblock}
+template <InputIterator I, Sentinel<I> S, NoThrowForwardIterator O>
+  requires Constructible<value_type_t<O>, reference_t<I>>() &&
+           __ReferenceTo<O, value_type_t<O>>()
+    tagged_pair<tag::in(I), tag::out(O)>
+    uninitialized_copy(I first, S last, O result);
+\end{codeblock}
+} %% \color{addclr}
+
+\setcounter{Paras}{0}
+\pnum
+\effects \changed{As if by}{Equivalent to}:
+\begin{codeblock}
+        for (; first != last; ++result, (void)++first)
+          ::new @\added{(const_cast<void*>}@(static_cast<@\added{const volatile }@void*>(addressof(*result)))@\added{)}@
+            @\changed{typename iterator_traits<ForwardIterator>::value_type}{value_type_t<O>}@(*first);
+        @\added{return \{first, result\};}@
+\end{codeblock}
+
+{\color{addclr}
+\requires \tcode{result} shall not be in the range \tcode{[first, last)}.
+
+\begin{codeblock}
+template <InputRange Rng, NoThrowForwardIterator O>
+  requires Constructible<value_type_t<O>, reference_t<iterator_t<Rng>>>() &&
+           __ReferenceTo<O, value_type_t<O>>()
+    tagged_pair<tag::in(safe_iterator_t<Rng>), tag::out(O)>
+    uninitialized_copy(Rng&& rng, O result);
+\end{codeblock}
+
+\pnum
+\effects Equivalent to:
+\begin{codeblock}
+        return uninitialized_copy(begin(rng), end(rng), result);
+\end{codeblock}
+
+\pnum
+\requires \tcode{result} shall not be in the range \tcode{[begin(rng), end(rng))}.
+
+\begin{codeblock}
+template <InputIterator I, NoThrowForwardIterator O>
+  requires Constructible<value_type_t<O>, reference_t<I>>() &&
+           __ReferenceTo<O, value_type_t<O>>()
+    tagged_pair<tag::in(I), tag::out(O)>
+    uninitialized_copy_n(I first, difference_type_t<I> n, O result);
+\end{codeblock}
+
+\pnum
+\effects Equivalent to:
+\begin{codeblock}
+        return uninitialized_copy(make_counted_iterator(first, n),
+                                  default_sentinel{}).base();
```

Comment:
This needs to be a little more complicated: `uninitialized_copy` returns a `tagged_pair` of `counted_iterator<I>` and `O` that we need to partially unwrap:
```c++
auto foo = uninitialized_copy(make_counted_iterator(first, n), default_sentinel{});
return {foo.in().base(), foo.out()};
```